processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinable.java

```diff
@@ -97,41 +103,49 @@ public JoinMatcher makeJoinMatcher(
+        IndexedTable.Reader reader = table.columnReader(correlatedColumnPosition);
+        closer.register(reader);
+        IntList rowIndex = index.find(searchColumnValue);
+        for (int i = 0; i < rowIndex.size(); i++) {
+          int rowNum = rowIndex.getInt(i);
+          String correlatedDimVal = Objects.toString(reader.read(rowNum), null);
+          correlatedValues.add(correlatedDimVal);
 
-    Set<String> correlatedValues = new HashSet<>();
-    if (table.keyColumns().contains(searchColumnName)) {
-      IndexedTable.Index index = table.columnIndex(filterColumnPosition);
-      IndexedTable.Reader reader = table.columnReader(correlatedColumnPosition);
-      IntList rowIndex = index.find(searchColumnValue);
-      for (int i = 0; i < rowIndex.size(); i++) {
-        int rowNum = rowIndex.getInt(i);
-        String correlatedDimVal = Objects.toString(reader.read(rowNum), null);
-        correlatedValues.add(correlatedDimVal);
-
-        if (correlatedValues.size() > maxCorrelationSetSize) {
+          if (correlatedValues.size() > maxCorrelationSetSize) {
+            return Optional.empty();
+          }
+        }
+        return Optional.of(correlatedValues);
+      } else {
+        if (!allowNonKeyColumnSearch) {
           return Optional.empty();
         }
-      }
-      return Optional.of(correlatedValues);
-    } else {
-      if (!allowNonKeyColumnSearch) {
-        return Optional.empty();
-      }
 
-      IndexedTable.Reader dimNameReader = table.columnReader(filterColumnPosition);
-      IndexedTable.Reader correlatedColumnReader = table.columnReader(correlatedColumnPosition);
-      for (int i = 0; i < table.numRows(); i++) {
-        String dimVal = Objects.toString(dimNameReader.read(i), null);
-        if (searchColumnValue.equals(dimVal)) {
-          String correlatedDimVal = Objects.toString(correlatedColumnReader.read(i), null);
-          correlatedValues.add(correlatedDimVal);
-          if (correlatedValues.size() > maxCorrelationSetSize) {
-            return Optional.empty();
+        IndexedTable.Reader dimNameReader = table.columnReader(filterColumnPosition);
+        IndexedTable.Reader correlatedColumnReader = table.columnReader(correlatedColumnPosition);
+        closer.register(dimNameReader);
+        closer.register(correlatedColumnReader);
+        for (int i = 0; i < table.numRows(); i++) {
+          String dimVal = Objects.toString(dimNameReader.read(i), null);
+          if (searchColumnValue.equals(dimVal)) {
+            String correlatedDimVal = Objects.toString(correlatedColumnReader.read(i), null);
+            correlatedValues.add(correlatedDimVal);
+            if (correlatedValues.size() > maxCorrelationSetSize) {
+              return Optional.empty();
+            }
           }
         }
-      }
 
-      return Optional.of(correlatedValues);
+        return Optional.of(correlatedValues);
+      }
+    }
+    finally {
+      CloseQuietly.close(closer);
```

Comment:
Can we use try-with-resources instead of `CloseQuiently`? It would be more aligned with https://github.com/apache/druid/pull/10247 as well.

Reply:
removed all usage of `CloseQuietly` in this PR